codec/unknownproto/unknown_fields.go

```diff
@@ -62,19 +69,27 @@ func (ckr *Checker) RejectUnknownFields(b []byte, msg proto.Message) error {
 			}
 
 		default:
-			if !ckr.AllowUnknownNonCriticals || tagNum&bit11NonCritical == 0 {
+			isCriticalField := tagNum&bit11NonCritical == 0
+
+			if !isCriticalField {
+				hasUnknownNonCriticals = true
+			}
+
+			if isCriticalField || !allowUnknownNonCriticals {
 				// The tag is critical, so report it.
-				return &errUnknownField{
+				return hasUnknownNonCriticals, &errUnknownField{
 					Type:     reflect.ValueOf(msg).Type().String(),
 					TagNum:   tagNum,
 					WireType: wireType,
 				}
 			}
 		}
 
-		// Skip over the 2 bytes that store fieldNumber and wireType bytes.
-		fieldBytes := b[2:n]
-		b = b[n:]
+		// Skip over the bytes that store fieldNumber and wireType bytes.
+		bz = bz[m:]
+		n := protowire.ConsumeFieldValue(tagNum, wireType, bz)
+		fieldBytes := bz[:n]
```

Comment:
In what cases don't we have 2 bytes? The protobuf specification stores the field number and wireType ALWAYS in 2 bytes. Also ConsumeField does what you've inlined. Perhaps could you explain this change?

Reply:
That's not true. Field num is a varint. The previous version didn't follow the proto spec.

Reply:
More specifically as I believe the two bytes consumed before actually covered the case where the field number was in the range 1-15 (byte 1) and the nested message was 0-127 bytes long (byte 2). I believe we were just missing test cases for messages outside of those criteria. So at some point as a followup, I'd like to add even more tests.

Reply:
Interesting, thank you for pointing that out; I'll revise my understanding and restudy the encoding. Also I believe I had those nested message tests already.

Reply:
The existing tests probably didn't trigger a condition where it wasn't 2 bytes unfortunately. I did come across those cases in other parts of the code.